migrator/migrations/m_192_to_m_193_report_config_migrations/migration_impl.go

```diff
@@ -0,0 +1,236 @@
+			if err != nil {
+				return errors.Wrapf(err, "failed to convert report config from proto %+v", reportConfigProto)
+			}
+			convertedReportConfigs = append(convertedReportConfigs, convertedGormConfig)
+
+			continue
+		}
+
+		// create v2 report config from v1
+		newConfig := createV2reportConfig(reportConfigProto)
+
+		// create notifier config for report_configuration_notifier table
+		notifierConfig := createNotifier(reportConfigProto)
+		newConfig.Notifiers = append(newConfig.Notifiers, notifierConfig)
+
+		// add notifier to notifier_configurations_notifiers
+		reportNotifierGorm, err := updatedSchema.ConvertNotifierConfigurationFromProto(notifierConfig, 0, newConfig.GetId())
+		if err != nil {
+			errMsg := errors.Wrapf(err, "failed to convert notifier config from proto %+v", notifierConfig)
+			errList.AddError(errMsg)
+		}
+		reportNotifiers = append(reportNotifiers, reportNotifierGorm)
+		// assign version to 2 to new copy and version to 1 in original so that they are not re-created during migration
+		newConfig.Version = 2
+		reportConfigProto.Version = 1
+		// convert report config proto back to gorm model
+		convertedGormNewConfig, err := updatedSchema.ConvertReportConfigurationFromProto(newConfig)
+		if err != nil {
+			return errors.Wrapf(err, "failed to convert report config from proto %+v", newConfig)
+		}
+		convertedGormReportConfig, err := updatedSchema.ConvertReportConfigurationFromProto(reportConfigProto)
+		if err != nil {
+			errMsg := errors.Wrapf(err, "failed to convert from proto %+v", reportConfigProto)
+			errList.AddError(errMsg)
+		}
+		convertedReportConfigs = append(convertedReportConfigs, convertedGormNewConfig, convertedGormReportConfig)
+
+		// create report snapshot for last run report job
+		reportSnapshot := createReportSnapshot(reportConfigProto, newConfig)
+		if reportSnapshot != nil {
+			// convert report snapshot to GORM
+			reportSnapshotGORM, err := updatedSchema.ConvertReportSnapshotFromProto(reportSnapshot)
+			if err != nil {
+				errMsg := errors.Wrapf(err, "failed to convert report snapshot from proto %+v", reportSnapshot)
+				errList.AddError(errMsg)
+			}
+			reportSnapshots = append(reportSnapshots, reportSnapshotGORM)
+		}
+		if len(convertedReportConfigs) == batchSize {
+			err = updateTablesTransaction(tx, convertedReportConfigs, reportSnapshots, reportNotifiers)
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
+			convertedReportConfigs = convertedReportConfigs[:0]
+			reportSnapshots = reportSnapshots[:0]
+			reportNotifiers = reportNotifiers[:0]
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if rows.Err() != nil {
+		return errors.Wrap(rows.Err(), "failed to get rows for report_configurations")
+	}
+
+	err = updateTablesTransaction(tx, convertedReportConfigs, reportSnapshots, reportNotifiers)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	if !errList.Empty() {
+		log.Error(errList)
+	}
```

Comment:
`errList` is holding errors from conversion, which should be checked before updating the tables. This migration has its challenges. If the report config conversion fails, but the notifier or snapshot conversion succeeds, the update table transaction will fail due to foreign key constraint violation.

You have to make sure that you remove the config, notifier, and snapshot from the respective array if at least one failed to convert. You can possibly use a map of config -> struct{[]notifiers, snapshot} to tackle the issue.